lib/burnside/intrinsics.cc

```diff
@@ -0,0 +1,99 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "intrinsics.h"
+#include "builder.h"
+
+/// [Coding style](https://llvm.org/docs/CodingStandards.html)
+
+namespace Fortran::burnside {
+
+// Define a simple static runtime description that will be transformed into
+// IntrinsicImplementation when building the IntrinsicLibrary.
+namespace runtime {
+enum class Type { f32, f64 };
+struct StaticDescription {
+  const char *name;
+  const char *symbol;
+  Type resultType;
+  std::vector<Type> argumentTypes;
+};
+
+// TODO
+// This table should be generated in a clever ways and probably shared with
+// lib/evaluate intrinsic folding.
+static const StaticDescription llvm[] = {
+    {"abs", "llvm.fabs.f32", Type::f32, {Type::f32}},
+    {"abs", "llvm.fabs.f64", Type::f64, {Type::f64}},
+    {"acos", "acosf", Type::f32, {Type::f32}},
+    {"acos", "acos", Type::f64, {Type::f64}},
+    {"atan", "atan2f", Type::f32, {Type::f32, Type::f32}},
+    {"atan", "atan2", Type::f64, {Type::f64, Type::f64}},
+    {"sqrt", "llvm.sqrt.f32", Type::f32, {Type::f32}},
+    {"sqrt", "llvm.sqrt.f64", Type::f64, {Type::f64}},
+    {"cos", "llvm.cos.f32", Type::f32, {Type::f32}},
+    {"cos", "llvm.cos.f64", Type::f64, {Type::f64}},
+    {"sin", "llvm.sin.f32", Type::f32, {Type::f32}},
+    {"sin", "llvm.sin.f64", Type::f64, {Type::f64}},
+};
+
+// Conversion between types of the static representation and MLIR types.
+mlir::Type toMLIRType(Type t, mlir::MLIRContext &context) {
+  switch (t) {
+  case Type::f32: return mlir::FloatType::getF32(&context);
+  case Type::f64: return mlir::FloatType::getF64(&context);
+  }
+}
+mlir::FunctionType toMLIRFunctionType(
+    const StaticDescription &func, mlir::MLIRContext &context) {
+  std::vector<mlir::Type> argMLIRTypes;
+  for (runtime::Type t : func.argumentTypes) {
+    argMLIRTypes.push_back(toMLIRType(t, context));
+  }
+  return mlir::FunctionType::get(
+      argMLIRTypes, toMLIRType(func.resultType, context), &context);
+}
+}  // runtime
+
+std::optional<mlir::FuncOp> IntrinsicLibrary::getFunction(
+    const std::string &name, const mlir::Type &type,
+    mlir::OpBuilder &builder) const {
+  auto module{getModule(&builder)};
+  if (const auto &it{lib.find({name, type})}; it != lib.end()) {
+    const IntrinsicImplementation &impl{it->second};
+    if (mlir::FuncOp func{getNamedFunction(impl.symbol)}) {
+      return func;
+    }
+    mlir::FuncOp function{createFunction(module, impl.symbol, impl.type)};
+    function.setAttr("fir.intrinsic", builder.getUnitAttr());
+    return function;
+  } else {
+    return std::nullopt;
+  }
+}
+
+// So far ignore the version an only load the dummy llvm lib.
+IntrinsicLibrary IntrinsicLibrary::create(
+    IntrinsicLibrary::Version, mlir::MLIRContext &context) {
```

Comment:
Just a thought. Would it work to statically instantiate the table for all versions and select the right one at runtime?

Reply:
Not in its current format because it stores mlir::Type that I only know hoe to build dynamically because it requires the "dynamic". `mlir::MLIRContext`.
We could keep the static table and perform the search directly with it, but that would be slower for the look ups (I think) and I also do not want to propagate the "Type" enum stuff at the interface.

Reply:
Kept as it is.